Model configuration output in HDF

andrewfullard opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Unit test added for a `to_hdf` method of the configuration namespace. Successfully converts a configuration to a YAML-formatted string.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Required to output simulation input parameters as part of the HDF in the standard HDF output from simulations.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Unit test added and fails

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have assigned/requested two reviewers for this pull request.

andrewfullard committed:
Added test for config -> HDF

andrewfullard committed:
Unit test moved and HDFMixin subclassed

Unit test moved to the config reader tests and the Configuration class
is now subclassed to with ConfigWriterMixin, itself a subclass of
HDFWriterMixin. get_properties has been overridden to obtain properties
from configuration dict keys.

andrewfullard committed:
Improved test and config reading

Config is now returned as a dict of dicts instead of a dict containing configurationnamespace objects. Test doesn't need the pytest fixture any more.

andrewfullard committed:
Mostly valid HDF is now written out

Test still fails, but the HDF contains configuration information in CGS units. Scalar output means that some entries are not well formatted. Pandas still cannot read the HDF.

andrewfullard commented:
Convert ConfigurationNameSpace to YAML and then dump to string instead?

andrewfullard committed:
Test passes, config written to string

Rewrote ConfigWriterMixin, moved it to util.py to match PlasmaWriterMixin, added static method override for to_hdf_util to handle the string output, updated test to do string-to-string comparison

andrewfullard committed:
Cleanup of unused imports

wkerzendorf commented on tardis/io/util.py:
```diff
@@ -399,3 +399,75 @@ def download_from_url(url, dst):
                 pbar.update(1024)
     pbar.close()
     return file_size
+
+
```
```suggestion

def quantity_representer(dumper, data):
    """
    docu goes here
    """
    return dumper.represent_data(str(data))
def cns_representer(dumper, data):
    """
    docu goes here
    """
    return dumper.represent_dict(dict(data))
```

wkerzendorf commented:
`pd.Series({'config':x})`

andrewfullard committed:
Rewrite to use yaml representer

Simplifies ConfigWriterMixin using YAML representers to read Quantities and Configuration objects

wkerzendorf commented on tardis/io/tests/test_config_reader.py:
```diff
@@ -57,3 +58,13 @@ def test_from_config_dict(tardis_config_verysimple):
+            tardis_config_verysimple, validate=True, config_dirname="test"
+        )
+    expected.to_hdf(hdf_file_path)
+    actual = pd.read_hdf(hdf_file_path, key="/config/config")
```
why config/config?

andrewfullard replied:
Fixed in next commit

andrewfullard committed:
Moved the config to the simulation part of HDF
